### DISCUSSION TITLE
Error (-10) T-Beam & lorawan

### DISCUSSION THREAD

**Discussion by icicmoi** [jgromes/RadioLib#904] [Q&A]
******************************
Error in the tittle : error 12 and not 10
*****************************

Hi all,
I am still unable to configure a lora32 T-Beam T22 V1-1 module (Lilygo) which contains an SX1262.
I used this example code: _RadioLib/examples/LoRaWAN/LoRaWAN_End_Device_ABP/LoRaWAN_End_Device_ABP.ino_

but I get an error:
**[LoRaWAN] Sending uplink packet ... failed, code -12**
(in TypeDef.h I read :
The supplied frequency value is invalid for this module.
#define RADIOLIB_ERR_INVALID_FREQUENCY but I didn't do anything at frequency level other than enter the band.
( LoRaWANNode node(&radio, &EU868); )

Thank you very much for your advice !

ps1 :
I chose this setting:
SX1262 radio = new Module(**18, 26, 23, 33); // nss, dio0, reset, dio1**
based on those given here :
_https://github.com/Xinyuan-LilyGO/LilyGo-LoRa-Series/blob/master/schematic/LilyGo_TBeam_V1.1.pdf_

ps2 : Sorry for this probably obvious problem to solve  but I'm just a simple Lorawan user ... and I don't know which way to look :-(

**Comment by jgromes** [comment]
SX1262 should be able to use 868 MHz, the root cause will be elsewhere. I think it might be an unitialized peristent storage. Can you run the example with both debug and verbose? See https://github.com/jgromes/RadioLib/wiki/Debug-mode

**Reply by StevenCellist** [reply]
If one were to guess that I never checked if ABP still functions after the whole revamp, that would be a correct guess as ABP is not the way to go and I never use it. Will check out if my lack of attention to ABP is actually the cause.
But on the other hand, I haven't encountered the need to specifically call `node.wipe()` on my personal test modules.. it works fine just fine until now. So what the cause may be is yet to be seen. But at least I'll give ABP a shot.

**Comment by icicmoi** [comment]
Thanks a lot to both of you @jgromes and @StevenCellist for replying so quickly but I think I've reached my level of incompetence :-(
1/ When I choose to accept verbose and/or debug (in BuildOptUser.h) I don't see any more information in my serial monitor.
Only, and always, 
[SX1262] Initializing ... success!
[LoRaWAN] Attempting over-the-air activation ... success!
[LoRaWAN] Sending uplink packet ... failed, code -12

2/ In addition, and further to your comment 
> I think it might be an unitialized peristent storage

I ran this command :
esptool.py --port /dev/ttyUSB0 erase_flash
esptool.py v4.7.0
Serial port /dev/ttyUSB0
Connecting......
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting....
Detecting chip type... ESP32
Chip is ESP32-D0WDQ6-V3 (revision v3.0)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: 3c:61:05:0d:5d:fc
Uploading stub...
Running stub...
Stub running...
Erasing flash (this may take a while)...
Chip erase completed successfully in 6.3s
Hard resetting via RTS pin...

Useful or useless ... I don't know ... but no change observed after uploading the sketch again.

Thanks .
ps : ok @StevenCellist for the wrong choice of abp ... but it will be used temporarily so that we can get an idea of how to position our nodes - with the help of ttnmapper - as our little association only has one external gateway. 
I prefer abp (but I may be wrong) for such tests with movement and GPS positioning.

**Reply by StevenCellist** [reply]
Minor details all over the place, but if you have a day or two spare and some patience, I will spin up a device in ABP mode to see for myself if there are any bugs. I will let you know if it works for me or if there's something that I made a mistake in!

> I prefer abp (but I may be wrong) for such tests with movement and GPS positioning.

The 'only' difference between ABP and OTAA is that you need to join at the beginning of a session which requires a gateway in range, and if your board has persistent storage, that means only once ever ever ever unless you perform the erase_flash command, because the session is saved in memory and reused on next boot. After the initial join, there is not a 'single bit of difference' between ABP and OTAA! 
And as a bonus, you don't need to fill in as many keys in OTAA as for ABP.. less room for mistakes, yay :)

**Reply by icicmoi** [reply]
I'm more the stubborn type, so this requires patience !
So I don't have to worry about waiting until you're available; that's great, thank you.
(and as far as OTAA is concerned, it's because our code wasn't designed - on arduino pro mini - to store session keys that it didn't work!)
I've just moved up a rung on the "Lorawan" learning ladder !!!!

**Comment by StevenCellist** [comment]
```
[LoRaWAN] Sending uplink packet ... There are no channels defined - are you in ABP mode with no defined subband?

Channel frequency UL = 0.000000 MHz
failed, code -12
```
Problem in the code.. investigating. Should work out-of-the-box on bands like EU868 that have no subbands.

**Reply by StevenCellist** [reply]
Found the culprit - there's no part in the code that initializes the starting datarate! In OTAA this is done during the join procedure, but ABP of course doesn't have that :)
Applying a fix...

**Reply by StevenCellist** [reply]
Hm, lack of antenna here and too far from gateway. Code appears to be working, will test later today when I can attach an antenna.

**Reply by StevenCellist** [reply]
Fixed the problem, but I now see that 'activating' an ABP device (i.e. a first uplink) causes a very very very fat MAC downlink with a payload of 52 bytes pure MAC commands. Which is not handled correctly (unsurprisingly). I will investigate. Here's the info for the memes:

![image](https://github.com/jgromes/RadioLib/assets/47155822/5313aeac-fd30-4cb3-a332-b60b5a1897e0)

![image](https://github.com/jgromes/RadioLib/assets/47155822/902862e9-5a3f-4c90-bb13-3811ec7f91e7)

**Reply by StevenCellist** [reply]
@icicmoi due to Christmas being around the corner, I don't expect to be able to take a better look at this until somewhere around New Year. I recommend you to use OTAA instead which has been confirmed by many to be working well (and is the better option either way).
At the end of the first week of January, I hope to publish bugfixes that will solve these ABP problems along other improvements / additions.

**Reply by icicmoi** [reply]
no problem @StevenCellist : thank you so much for looking into my question so quickly !
Happy New Year to you all! .

**Comment by StevenCellist** [comment]
@icicmoi happy New Year to you too!

Meanwhile, there's significant progress:
![image](https://github.com/jgromes/RadioLib/assets/47155822/6f13542f-2d14-4677-8210-a9a88b4c64e0)
Most stuff is working and showing up; there are minor points to be solved still.
